### ISSUE TITLE
Strange non-deterministic gem version resolution

### ISSUE THREAD

**Issue by jhawthorn** [bundler/bundler#5633]
Working on [a gem](https://github.com/solidusio-contrib/solidus_product_feed) today I experienced very strange, incorrect, and non-deterministic version resolution. Please bear with me, this is a weird one, but I have a consistent reproduction at the end.

It started with this:

[![](http://i.hawth.ca/s/Z81aVYhF.png)](https://twitter.com/jhawthorn/status/859535068821245952)

Every time I ran `bundle update`, it would downgrade Rails by one minor version. From `5.0.2` to `5.0.1` to `5.0.0.1` and finally to `5.0.0`. [Here is a capture of my tmux pane as I was running these bundle updates](https://gist.github.com/jhawthorn/1778776a5a1c8cdd225342993db41dae).

I couldn't reproduce this because it eventually decided that it really liked rails 5.0.0.1, and decided to stick with that version. However I found then that running `bundle update rails` would bump me up to rails `5.0.2` (the version it should be) without downgrading anything else (implying that should have always been the correct version). Running `bundle update` after this would again downgrade rails to 5.0.0.1. Something strange is going on here.

I managed to get this down to a [minimal Gemfile](https://github.com/jhawthorn/bundler_wtf/blob/master/Gemfile) which exhibits the issue:
```
source "https://rubygems.org"
gem 'solidus_core', '2.2.0'
gem 'rspec-rails', '~> 3.4'
```

You can see this in a terminal capture I have made here: http://showterm.io/917485215a8753019f0e6

Thank you for reading this far :heart: , I know "my bundler downgrades rails every time I run it" is an extraordinary claim. I hope I've provided the evidence to somewhat back it up.

Here is my `bundler env` (bundler `1.14.6`, ruby `2.3.3`) for
* After `bundle update` [with rails 5.0.0.1](https://gist.github.com/2cfeab028c24d8d88322603a89b05e4e)
* After `bundle update rails` [with rails 5.0.2](https://gist.github.com/c92fe8bdaeed5921e7795734c76a6ca4)

**Comment by stewart** [comment]
It appears this is a v1.14.0 regression, as I'm able to recreate it there but not in v1.13.7.

**Comment by segiddins** [comment]
Is this fixed in 1.15.0.pre.3, perchance?

**Comment by jhawthorn** [comment]
Same issue in 1.15.0.pre.3

**Comment by segiddins** [comment]
Damn, I know I fixed another `bundle update` bug in 1.15 and was hoping they'd be the same :(
I can't promise a swift resolution to this bug, so in the meantime I recommend giving your CPU (and sanity) a break by pinning a rails version in your Gemfile

**Comment by segiddins** [comment]
```diff
diff --git a/spec/realworld/edgecases_spec.rb b/spec/realworld/edgecases_spec.rb
index 302fd57cf..6a4cc87e1 100644
--- a/spec/realworld/edgecases_spec.rb
+++ b/spec/realworld/edgecases_spec.rb
@@ -94,6 +94,21 @@ RSpec.describe "real world edgecases", :realworld => true, :sometimes => true do
     expect(lockfile).to include(rubygems_version("activesupport", "~> 3.0"))
   end
 
+  it "i hate everything", :sometimes => false do
+    gemfile <<-G
+      source "https://rubygems.org"
+
+      gem "solidus_core", "2.2.0"
+      gem "rspec-rails", "~> 3.4"
+    G
+
+    bundle! :lock
+    expect(the_bundle.locked_gems.specs.map(&:full_name)).to include("rails-5.0.2")
+
+    bundle! "lock --update"
+    expect(the_bundle.locked_gems.specs.map(&:full_name)).to include("rails-5.0.2")
+  end
+
   it "is able to update a top-level dependency when there is a conflict on a shared transitive child", :ruby => "2.1" do
     # from https://github.com/bundler/bundler/issues/5031
 
```

**Comment by segiddins** [comment]
Making this even harder to track down is that this appears to be a truly pathological gemfile, it breaks in different ways on different versions, rendering `git bisect` kinda useless (see v1.14.0.pre.1 where there's an unrelated failure)

**Comment by segiddins** [comment]
It seems we broke this in f3836b6a1ec0620d164e52b8afec830b13494858

**Comment by segiddins** [comment]
Disabling swapping in the resolver fixes things. 🔥

**Comment by jhawthorn** [comment]
Unfortunately I am unable to specify a version of rails or use a lock file because I am developing gems.

Something has changed in the last week (maybe more versions of rails released?) and this is now far more broken far more often. About a dozen gems I maintain now cannot bundle correctly (admittedly, they will have gemfiles very similar to this one).

I'm also now receiving this error (on bundler `1.15.0.pre.4`), which looks very wrong.

```
Bundler could not find compatible versions for gem "actionpack":
  In Gemfile:
    solidus was resolved to 2.3.0.alpha, which depends on
      solidus_core (= 2.3.0.alpha) was resolved to 2.3.0.alpha, which depends on
        rails (~> 5.0.0) was resolved to 5.0.0, which depends on
          actionpack (= 5.1.1)

    solidus was resolved to 2.3.0.alpha, which depends on
      solidus_core (= 2.3.0.alpha) was resolved to 2.3.0.alpha, which depends on
        rails (~> 5.0.0) was resolved to 5.0.0, which depends on
          actionpack (= 5.1.1)

    solidus was resolved to 2.3.0.alpha, which depends on
      solidus_core (= 2.3.0.alpha) was resolved to 2.3.0.alpha, which depends on
        rails (~> 5.0.0) was resolved to 5.0.0, which depends on
          actionpack (= 5.0.0)
```

Bundler seems to think that `rails (=5.0.0)` depends on `actionpack (= 5.1.1)` :thinking:

**Comment by jhawthorn** [comment]
You can reproduce that error right now with the Gemfile:

```
source "https://rubygems.org"

gem 'solidus_core', '2.2.0'
gem 'rails', '>= 5.0.0' # This line added since last time

gem 'rspec-rails', '~> 3.4'
```

**Comment by jhawthorn** [comment]
Here's the output of `bundle update` with `DEBUG_RESOLVER=1` https://gist.github.com/jhawthorn/1fc65fca917414dacaea3110ebd96a59

**Comment by segiddins** [comment]
This should be fixed by https://github.com/CocoaPods/Molinillo/pull/69

**Comment by jhawthorn** [comment]
This does seem to be fixed on bundler `1.16.0.pre.1` :tada:

**Comment by segiddins** [comment]
Yay!
Great job @greysteil
